segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,500 @@
+                iterationDelay.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * This function instantiates a new Segment supplier by creating actual Segment against the {@link SegmentContainer}
+     * passed as input based on the name passed to the function. Note that the Segment is only created if it does not
+     * exists. Otherwise, it is just loaded and returned.
+     *
+     * @param container {@link SegmentContainer} to create Segments from.
+     * @param timeout   Timeout for the Segment creation to complete.
+     * @return A future that, when completed, contains reference to the Segment to be used by a given
+     * {@link ContainerEventProcessor.EventProcessor} based on its name.
+     */
+    private static Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
+                                                                                                       MetadataStore metadataStore,
+                                                                                                       Duration timeout) {
+        return s -> Futures.exceptionallyComposeExpecting(
+                container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout),
+                e -> e instanceof StreamSegmentNotExistsException,
+                () -> metadataStore.submitAssignment(newSegment(getEventProcessorSegmentName(container.getId(), s),
+                        SYSTEM_CRITICAL_SEGMENT, Collections.emptyList()), true, timeout) // Segment should be pinned.
+                        .thenCompose(l -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout)));
+    }
+
+    /**
+     * Reports the metrics periodically for each {@link ContainerEventProcessor.EventProcessor}.
+     */
+    private void reportMetrics() {
+        for (EventProcessorImpl ep : eventProcessorMap.values()) {
+            SegmentStoreMetrics.outstandingEventProcessorBytes(ep.getName(), containerId, ep.getOutstandingBytes());
+            // Only report the last iteration processing latency if it has been successful.
+            if (!ep.failedIteration.get()) {
+                ep.metrics.batchProcessingLatency(ep.lastIterationLatency.get());
+            }
+        }
+    }
+
+    //endregion
+
+    //region AutoCloseable Implementation
+
+    @Override
+    public void close() {
+        if (!closed.getAndSet(true)) {
+            log.info("{}: Closing ContainerEventProcessor service.", this.traceObjectId);
+            closeProcessors();
+            eventProcessorMetricsReporting.cancel(true);
+        }
+    }
+
+    private void closeProcessors() {
+        synchronized (eventProcessorMap) {
+            if (eventProcessorMap.isEmpty()) {
+                return;
+            }
+            eventProcessorMap.forEach((k, v) -> {
```

Comment:
You are looping through `eventProcessorMap`, yet in your `EventProcessor.close` method you invoke a `remove` from this map. This should throw a `ConcurrentModificationException` - how come this works?

That is, for most of these collections, modifying the collection while iterating through it is disallowed. 

Can you please check that this works as intended?

Reply:
This method is safe as we are using a `ConcurrentHashMap` to store `EventProcessor` references. I have added a new test to validate that (I have also verified that using a `HashMap` in `ContainerEventProcessorImpl` instead of `ConcurrentHashMap` leads to `ConcurrentModificationException` in that new test).